src/SourceGenerators/Uno.UI.SourceGenerators/HotRestart/HotRestartGenerator.cs

```diff
@@ -0,0 +1,352 @@
+				{
+					return;
+				}
+
+				_mauiAppDelegate = _context.Compilation.GetTypeByMetadataName("Microsoft.Maui.MauiUIApplicationDelegate");
+
+				var builder = new IndentedStringBuilder();
+				builder.AppendLineIndented("#if DEBUG");
+				builder.AppendLineIndented("// <auto-generated>");
+				builder.AppendLineIndented("// *************************************************************");
+				builder.AppendLineIndented($"// This file has been generated by Uno.UI ({nameof(HotRestartGenerator)})");
+				builder.AppendLineIndented("// *************************************************************");
+				builder.AppendLineIndented("// </auto-generated>");
+				builder.AppendLine();
+				builder.AppendLineIndented("using System;");
+				builder.AppendLine();
+
+				TryGenerateMAUIAppDelegate(builder);
+				GenerateHotRestartDelegate(builder);
+				builder.AppendLineIndented("#endif");
+
+				_context.AddSource("HotRestartHelper.g.cs", builder.ToString());
+			}
+
+			private bool IsGenerationEnabled()
+				=> bool.TryParse(_context.GetMSBuildPropertyValue("UnoDisableHotRestartHelperGeneration", "false"), out var result) ? result : true;
+
+			private void GenerateHotRestartDelegate(IndentedStringBuilder builder)
+			{
+				builder.AppendLineIndented("[global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]");
+				using (builder.BlockInvariant("internal class __UnoHotRestartDelegate : global::Microsoft.Maui.MauiUIApplicationDelegate"))
+				{
+					using (builder.BlockInvariant("public override bool FinishedLaunching(global::UIKit.UIApplication uiApplication, global::Foundation.NSDictionary launchOptions)"))
+					{
+						var appType = GetApplicationDefinitionType();
+
+						builder.AppendLineIndented($"#if HAS_UNO_WINUI");
+						builder.AppendLineIndented($"Microsoft.UI.Xaml.Application.Start(_ => new {appType}());");
+						builder.AppendLineIndented($"#else");
+						builder.AppendLineIndented($"Windows.UI.Xaml.Application.Start(_ => new {appType}());");
+						builder.AppendLineIndented($"#endif");
+						builder.AppendLineIndented("return base.FinishedLaunching(uiApplication, launchOptions);");
+					}
+				}
+			}
+
+			private string GetApplicationDefinitionType()
+			{
+				var appDefinitionFile = _context.GetMSBuildItemsWithAdditionalFiles("ApplicationDefinition").FirstOrDefault();
+
+				if (appDefinitionFile?.File.GetText()?.ToString() is { } text)
+				{
+					// Read the xml file to get the x:Class attribute on the first node
+					var appDefinitionXml = System.Xml.Linq.XDocument.Parse(text);
+					var xClassAttribute = appDefinitionXml.Root?.Attributes().FirstOrDefault(a => a.Name.LocalName == "Class");
+
+					if (xClassAttribute is not null)
+					{
+						return "global::" + xClassAttribute.Value;
+					}
+					else
+					{
+						throw new InvalidOperationException($"Unable to find an x:Class attribute on the first node of the ApplicationDefinition");
```

Comment:
We shouldn't throw exceptions. I think it's better to use `_context.ReportDiagnostic`